Add export as HAR format for single requests

beardeddragon5 opened the pull request:
<!--
Please open an [Issue](https://github.com/kong/insomnia/issues/new) first to discuss new
features or non-trivial changes. Please provide as much detail as possible on the change as
possible including general description, implementation details, potential shortcommings, etc.

If this PR closes an issue, please mention "" where #XX is the issue number.
-->
Closes #650

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/Kong/insomnia?pullRequest=2715) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/Kong/insomnia?pullRequest=2715) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/Kong/insomnia?pullRequest=2715) it.</sub>

dimitropoulos committed:
rebase

dimitropoulos committed:
fixes errors

and adds some consistency while we're at it

dimitropoulos committed:
updates copy

dimitropoulos committed:
fixes type error in tests.

I was on my way to writing some tests to cover the things addressed in this PR.  I was happy to see that we already covered what I was planning to test.  However, there were some very surface-level type errors in the tests so I thought it appropriate to fix them while I was in this area of the code for this job.

develohpanda commented on packages/insomnia-app/app/common/har.ts:
```diff
@@ -172,6 +173,33 @@ export interface Har {
+    models.workspace.type,
+    models.requestGroup.type,
+  ]);
+  const workspace = ancestors.find(ancestor => ancestor.type === models.workspace.type);
```
We can use `isWorkspace` here

```suggestion
  const workspace = ancestors.find(isWorkspace);
```

dimitropoulos replied:
good one 7747354e58cfc50131dbf581228aaf4eac945a89

develohpanda commented on packages/insomnia-app/app/ui/components/panes/response-pane.tsx:
```diff
@@ -178,6 +178,41 @@ class ResponsePane extends PureComponent<Props> {
+  async _handleExportAsHAR() {
+    const { response, request } = this.props;
+
+    if (!response) {
+      // Should never happen
+      console.warn('No response to download');
+      return;
+    }
```
It's possible to export _without_ a response according to the `exportHar` implementation (although I know a response pane doesn't show when there's no response)

I think this is fine but it also seems fitting to add exporting a single request as HAR in the request context menu

<img width="269" alt="image" src="https://user-images.githubusercontent.com/4312346/134267692-9b33c4cb-d4b4-4781-9664-cc4c4846b9b6.png">

dimitropoulos replied:
@wdawson take a think about the above suggestion - we can always iterate.  seems ok to me, but I'm not sure how often people use HAR

wdawson replied:
From my experience, HAR is used when there is some sequence of requests/responses. I would rather not crowd this menu unless we find that there's a big demand for it there.

develohpanda commented on packages/insomnia-app/app/common/har.ts:
```diff
@@ -8,12 +8,13 @@ import * as models from '../models';
 import type { Request } from '../models/request';
 import { newBodyRaw } from '../models/request';
-import type { Response as ResponseModel } from '../models/response';
+import type { Response, Response as ResponseModel } from '../models/response';
```
Intentional aliased import for the same thing?

dimitropoulos replied:
nope, that alias was there from before, but we can remove it now.  I wasn't aware of the alias because I wasn't looking at the code that used it.

good find b8eae1f95c49a3dfa45a147b219a32b83a23ca3d

develohpanda reviewed (approved):
Seems to be working as intended!

<img width="879" alt="image" src="https://user-images.githubusercontent.com/4312346/134268003-f5957720-82e2-42a7-82b5-14a801ac3096.png">

dimitropoulos committed:
use isWorkspace selector

dimitropoulos committed:
fixes unnecessary alias

gatzjames commented on packages/insomnia-app/app/common/har.ts:
```diff
@@ -172,6 +174,33 @@ export interface Har {
+
+  const workspaceMeta = await models.workspaceMeta.getByParentId(workspace._id);
+  let environmentId = workspaceMeta ? workspaceMeta.activeEnvironmentId : null;
+  const environment = await models.environment.getById(environmentId || 'n/a');
```
Curious why is `n/a` needed here? Is it a pattern we use often or specific to this method?

develohpanda replied:
It's just a remnant of the past, especially when flow was a thing

gatzjames reviewed (approved):
🚀 Works great!

dimitropoulos committed:
Merge branch 'develop' into export-request-as-har
